File: bore.php
<?php

date_default_timezone_set('UTC');

$link = isset($_GET["link"]) ? $_GET["link"] : "2015052812/hires/needboreplot_2015052812_0305_0153_0239_0151_111_013.png";
$reqTime = isset($_GET["reqTime"]) ? $_GET["reqTime"] : "";

$url = "http://weather.ou.edu/~map/real_time_data/PECAN/PHI/".$link;

$s = explode("/",$link);
$localFile = "bore/".substr($s[2],0,-3)."txt";
$data = file($localFile);
$pieces = preg_split('/\s+/', $data[0]);

$invHeight = $pieces[1];
$invStrength = $pieces[2];
$cpDepth = $pieces[3];
$cpSpeed = $pieces[4];

/*
if($invHeight == -1){
	$invHeight = "";
}
if($invStrength == -1){
       	$invStrength = "";
}
if($cpDepth == -1){
       	$cpDepth = "";
}
if($cpSpeed == -1){
       	$cpSpeed = "";
}
*/

?>

<html>
<head>

<script type="text/javascript">

var timeint;
var boreURL;
var reqTime = <?php echo $reqTime; ?>;

function regenBore(){

	alert('Regenerating Bore Diagnostics, please stand by (image will automatically update).');
	document.getElementById('boreImg').src = '';

	var invHeight = document.getElementById('invHeight').value;
	var invStrength = document.getElementById('invStrength').value;
	var cpDepth = document.getElementById('cpDepth').value;
	var cpSpeed = document.getElementById('cpSpeed').value;

	var link = 'regenBore.php?reqTime=' + reqTime + '&invHeight=' + invHeight + '&invStrength=' + invStrength + '&cpDepth=' + cpDepth + '&cpSpeed=' + cpSpeed;
	var dataRequest = new XMLHttpRequest();
        dataRequest.open('GET', link, false);
        dataRequest.send();

	timerInt = setInterval(timer,1000);
}

function timer(){
	var request = new XMLHttpRequest();
        request.open('GET', 'bore/' + reqTime + '.txt', false);
        request.onreadystatechange = function(){
                if(request.readyState === 4){
                        if(request.status === 200 || request.status == 0){
				clearInterval(timerInt);
                                boreURL = 'http://weather.ou.edu/~map/real_time_data/PECAN/PHI/' + request.responseText.split("\n")[0];
				document.getElementById('boreImg').src = boreURL;
                        }
                }
        }
        request.send(null);
}

</script>

</head>
<body>

<table width="1878" border="1">
<tr>
<td width="150">

<b><u>Optional Parameter Modification</b>:</u>
<br>
<br>

<p>
Inversion Height (m):
<input type="text" id="invHeight" value="<?php echo $invHeight; ?>" />
</p>

<p>
Inversion Strength (K):
<input type="text" id="invStrength" value="<?php echo $invStrength; ?>" />
</p>

<p>
Cold Pool Depth (m):
<input type="text" id="cpDepth" value="<?php echo $cpDepth; ?>" />
</p>

<p>
Cold Pool Speed (m/s):
<input type="text" id="cpSpeed" value="<?php echo $cpSpeed; ?>" />
</p>

<p>
<input type="button" value="Submit" onClick="regenBore();" />
</p>

</td>
<td width="1728">

<img id="boreImg" src="<?php echo $url; ?>" width="1728" height="691" />

</td>
</tr>
</table>

</body>
</html>

File: bore.py
import os, sys
import numpy as np
import netCDF4
import urllib2
import time
import datetime

tNow = datetime.datetime.utcnow()

def findPoint(lons, lats, lon, lat):
        vals = np.abs(np.array(lons) - lon) + np.abs(np.array(lats) - lat)
        min = np.where(vals == vals.min())
        return min[0][0], min[1][0]

# get data from user
fh = open('/var/www/html/RadarQC/phi_new/bore/format.txt','r')
line = fh.readlines()[0].rstrip()
data = line.split(',')
fh.close()

# assign variables
# $str = $init.",".$x1.",".$y1.",".$x2.",".$y2.",".$member.",".$fHour;
initTime = data[0]
lon1 = float(data[1])
lat1 = float(data[2])
lon2 = float(data[3])
lat2 = float(data[4])
member = data[5]
fHour = data[6]
rTime = data[7]
invHeight = data[8]
invStrength = data[9]
cpDepth = data[10]
cpSpeed = data[11]

# get master lons/lats
if member == '111':
	nc = netCDF4.Dataset('/var/www/html/RadarQC/phi_new/2015060912_1km.nc')
else:
	nc = netCDF4.Dataset('/var/www/html/RadarQC/phi_new/2015052213_12.nc')
lats = nc.variables['XLAT'][:]
lons = nc.variables['XLON'][:]

# determine i, j points
py2, px2 = findPoint(lons, lats, lon1, lat1)
py1, px1 = findPoint(lons, lats, lon2, lat2)

# write out input string
# INITTIME X(COLDPOOL) Y(COLDPOOL) X(AMBIENT) Y(AMBIENT) VALIDTIME ENSSIZE MEMBER INITDATE
# 12 140 190 135 190 2015-05-16_16:00:00 20 1 20150515 
#      500       5    1000      10
#str = initTime + ' ' + str(px1) + ' ' + str(py1) + ' ' + str(px2) + ' ' + str(py2) + ' ' + validTime + ' ' + ensSize + ' ' + member + ' ' + initDate
st = invHeight.rjust(8) + invStrength.rjust(8) + cpDepth.rjust(8) + cpSpeed.rjust(8)
fName = 'needboreplot_' + initTime + '_' + str(px1).zfill(4) + '_' + str(py1).zfill(4) + '_' + str(px2).zfill(4) + '_' + str(py2).zfill(4) + '_' + member.zfill(3) + '_' + fHour.zfill(3) + '.txt'
iName = fName[:-3] + 'png'
fh = open('/var/www/html/RadarQC/phi_new/bore/' + fName,'w')
fh.write(st)
fh.close()

if member == 'hrrr':
	pass
	# fire off fortran code to extract sounding from the HRRR

	# fire off Kevin's code to generate bore diagnostics
else:
	# send to boomer
	cmd = 'scp -q /var/www/html/RadarQC/phi_new/bore/' + fName + ' kars0317@boomer.oscer.ou.edu:/work/pecanmap/boreinput/'
	print cmd
	os.system(cmd)

# build image URL
if member.zfill(3) == '111':
	mem = 'hires'
elif member.zfill(3) == '000':
	mem = 'ensmean'
else:
	mem = member.zfill(3)
link = initTime + '/' + mem + '/' + iName
url = 'http://weather.ou.edu/~map/real_time_data/PECAN/PHI/' + link
url2 = 'http://weather.ou.edu/~map/real_time_data/PECAN/PHI/' + initTime + '/' + mem

# wait for image to show up and then write notification to server
notHere = True
counter = 0
while notHere:
	try:
		f = urllib2.urlopen(urllib2.Request(url))

		f2 = urllib2.urlopen(url2)
		lines = f2.readlines()
		f2.close()

		for ln in lines:
			if iName in ln:
				s = ln.split()
				imgTime = datetime.datetime(int(s[7][:4]), int(s[7][5:7]), int(s[7][8:10]), int(s[8][:2]), int(s[8][3:5]))
				if imgTime >= tNow:
					fh = open('/var/www/html/RadarQC/phi_new/bore/' + rTime + '.txt','w')
			                fh.write(link + '\n' + line)
			                fh.close()

			                os.system('chmod 777 /var/www/html/RadarQC/phi_new/bore/' + rTime + '.txt')

			                notHere = False
				break
		if notHere:
			print 'modTime less that current time, waiting for image to update...'
			time.sleep(1)
			counter += 1
	                if counter > 120:
	                        sys.exit()

	except:
		print 'image not there yet, waiting...'
		time.sleep(1)
		counter += 1
		if counter > 120:
			sys.exit()


'''
print 'cross-section coords...'
print lons[96][122], lats[96][122], lons[218][122], lats[218][122]
print lons[96][213], lats[96][213], lons[218][213], lats[218][213]
print lons[96][288], lats[96][288], lons[218][288], lats[218][288]
print lons[96][122], lats[96][122], lons[96][288], lats[96][288]
print lons[146][122], lats[146][122], lons[146][288], lats[146][288]
print lons[218][122], lats[218][122], lons[218][288], lats[218][288]
'''
